etcdproxy/konnectivitycheck.go

```diff
@@ -0,0 +1,82 @@
+package etcdproxy
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type KonnectivityChecker struct {
+	metricsURL string
+	ready      chan struct{}
+	logger     *zap.Logger
+}
+
+func NewKonnectivityChecker(lg *zap.Logger, adminURL string, readyChan chan struct{}) *KonnectivityChecker {
+	return &KonnectivityChecker{
+		metricsURL: adminURL,
+		ready:      readyChan,
+		logger:     lg,
+	}
+}
+
+func (k *KonnectivityChecker) Run(c context.Context) error {
+	// First, wait until an agent has been registered.
+	err := wait.PollUntilContextCancel(c, 15*time.Second, true, func(ctx context.Context) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, nil
+		default:
+		}
+		resp, err := http.Get(k.metricsURL)
+		if err != nil {
+			k.logger.Error("failed to get metrics", zap.Error(err))
+			return false, nil
+		}
+		agentCount, err := getAgentCount(resp.Body)
+		if err != nil {
+			k.logger.Error("failed to get agent count", zap.Error(err))
+			return false, nil
+		}
+		if agentCount == 0 {
+			return false, nil
+		}
+		k.logger.Info("konnectivity agent registered", zap.Int("count", agentCount))
+		return true, nil
+	})
+	if err != nil {
+		k.logger.Error("failed to wait for konnectivity agent", zap.Error(err))
+		return err
+	}
+
+	// After the first agent has been registered, wait a fixed amount of time to allow more agents to register.
+	k.logger.Info("Waiting for more agents to register")
+	time.Sleep(90 * time.Second)
```

Comment:
Assuming this is just an arbitrary amount of time to wait or is it based on something?

Reply:
This is an arbitrary amount of time to wait. We can't know how many agents are going to join the konnectivity server (we know at least one because of the control plane agent). So I figured we'd give it a reasonable amount of time for others to join. In my testing, it never took more than 5s for others to join, so I figured 90s would be a conservative amount.